security.php

```diff
@@ -54,56 +55,148 @@ function wpcom_vip_is_restricted_username( $username ) {
+		 *
+		 * @param int    $threshold IP password reset threshold. Default 3.
+		 * @param string $ip        IP address of the password reset request.
+		 */
+		$ip_threshold = apply_filters( 'wpcom_vip_ip_password_reset_threshold', 3, $ip );
+
+		/**
+		 * Filters the threshold for limiting password resets by username.
+		 *
+		 * @param int    $threshold Username password reset threshold. Default is 5 times the IP + username combination threshold.
+		 * @param string $username  Username of the password reset request
+		 */
+		$username_threshold = apply_filters( 'wpcom_vip_username_password_reset_threshold', 5 * $ip_username_threshold, $username );
+	} elseif ( $is_restricted_username ) {
+		$ip_username_threshold = 2;
+	}
+
+	if ( $ip_username_count >= $ip_username_threshold || $ip_count >= $ip_threshold || $username_count >= $username_threshold ) {
+		$lock_interval = apply_filters( 'wpcom_vip_invalid_login_lock_period', MINUTE_IN_SECONDS * 30  );
+		if ( $cache_group === CACHE_GROUP_LOGIN_LIMIT ) {
+			do_action( 'login_limit_exceeded', $username );
+		} else {
+			do_action( 'password_reset_limit_exceeded', $username );
+		}
+
+		if ( $ip_username_count >= $ip_username_threshold ) {
+			wp_cache_set( CACHE_KEY_LOCK_PREFIX . $cache_keys['ip_username_cache_key'], true, $cache_group, $lock_interval );
+		} else if ( $ip_count >= $ip_threshold ) {
+			wp_cache_set( CACHE_KEY_LOCK_PREFIX . $cache_keys['ip_cache_key'], true, $cache_group, $lock_interval );
+		} else {
+			wp_cache_set( CACHE_KEY_LOCK_PREFIX . $cache_keys['username_cache_key'], true, $cache_group, $lock_interval );
+		}
+	}
+}
+
+
 /**
  * Tracks and caches IP, IP|Username events, and Username events.
  * We're tracking IP, and IP|Username events for both login attempts and
  * password resets but only tracking Username events for password resets.
  *
  * @param string $username The username to track.
  * @param string $cache_group The cache group to track the $username to.
- * @param int $cache_expiry The number in seconds of the cache expiry.
  */
-function wpcom_vip_track_auth_attempt( $username, $cache_group, $cache_expiry ) {
-	// phpcs:ignore WordPressVIPMinimum.Variables.ServerVariables.UserControlledHeaders
-	$ip                    = filter_var( $_SERVER['REMOTE_ADDR'] ?? '', FILTER_VALIDATE_IP, [ 'options' => [ 'default' => '' ] ] );
-	$ip_username_cache_key = $ip . '|' . $username; // IP + username
-	$ip_cache_key          = $ip; // IP only
-	$username_cache_key    = $username; // Username only
+function wpcom_vip_track_auth_attempt( $username, $cache_group ) {
+	$cache_keys = _wpcom_vip_login_cache_keys( $username );
+
+	$base_event_window = apply_filters( 'wpcom_vip_invalid_login_window', MINUTE_IN_SECONDS * 15  );
+	if ( $cache_group === CACHE_GROUP_LOST_PASSWORD_LIMIT ) {
+		$base_event_window = apply_filters( 'wpcom_vip_lost_password_window', MINUTE_IN_SECONDS * 30  );
+	}
 
 	// Longer TTL when logging in as admin, which we don't allow on WP.com
```

Comment:
Comment needs updating, since no VIP sites are on WP.com any more.

Reply:
Oooh does that mean we can drop that whole extended fail time for these usernames? I would rather not test it. At least not in the same PR. 

But if you know more about why this exception exists there I would be happy with extending the comment.

Reply:
We don't create sites with `admin` anyway, but the [function](https://github.com/Automattic/vip-go-mu-plugins/blob/b3ac61e7e73c7b33b24a82b7c2a3ab06715d6995/security.php#L51) also covers a couple of bot credentials too, so the exception is worth keeping in - you just need to update the comment to remove the reference to "WP.com".

Reply:
I changed the comment